x-pack/plugins/monitoring/server/kibana_monitoring/bulk_uploader.js

```diff
@@ -121,14 +121,35 @@ export class BulkUploader {
    * Bulk stats are transformed into a bulk upload format
    * Non-legacy transformation is done in CollectorSet.toApiStats
    */
-  static toBulkUploadFormat(uploadData) {
-    const payload = uploadData
-      .filter(d => Boolean(d) && !isEmpty(d.result))
-      .map(({ result, type }) => [{ index: { _type: type } }, result]);
-    if (payload.length > 0) {
-      const combinedData = BulkUploader.combineStatsLegacy(payload); // arrange the usage data into the stats
-      return flatten(combinedData);
+  toBulkUploadFormat(rawData, collectorSet) {
+    if (rawData.length === 0) {
+      return;
     }
+
+    // convert the raw data to a nested object by taking each payload through
+    // its formatter, organizing it per-type
+    const typesNested = rawData.reduce((accum, { type, result }) => {
```

Comment:
Can we include a sample of what the data structure looks like before and after each transformation? I find that helps understand intent a lot